agent/php_txn.c

```diff
@@ -536,6 +536,117 @@ static nrobj_t* nr_php_txn_get_labels() {
   return nr_labels_parse(NR_PHP_PROCESS_GLOBALS(env_labels));
 }
 
+static nr_status_t nr_php_txn_collect_label_keys_iter(const char* key,
+                                                      const nrobj_t* value,
+                                                      void* ptr) {
+  nrobj_t* user_data = (nrobj_t*)ptr;
+
+  (void)value;
+
+  if (NULL == user_data) {
+    return NR_FAILURE;
+  }
+
+  nro_set_array_string(user_data, 0, key);
+
+  return NR_SUCCESS;
+}
+
+/*
+ * Purpose : Filter the labels hash to exclude any labels that are in the
+ *           newrelic.application_logging.forwarding.labels.exclude list.
+ *
+ * Params  : 1. The labels hash to filter.
+ *
+ * Returns : A new hash containing the filtered labels.
+ *           If no labels exist or all labels are excluded, then return NULL.
+ *
+ */
+
+static nrobj_t* nr_php_txn_get_log_forwarding_labels(nrobj_t* labels) {
+  nrobj_t* label_keys = NULL;
+  nrobj_t* exclude_labels_list = NULL;
+  nrobj_t* exclude_labels_hash = NULL;
+  nrobj_t* log_labels = NULL;
+
+  if (NULL == labels || 0 == nro_getsize(labels)) {
+    nrl_verbosedebug(NRL_TXN, "%s: No labels defined", __FUNCTION__);
+    return NULL;
+  }
+
+  /* if logging labels are disabled then nothing to do */
+  if (0 == NRINI(log_forwarding_labels_enabled)) {
+    nrl_verbosedebug(NRL_TXN, "%s: Log forwarding labels disabled",
+                     __FUNCTION__);
+    return NULL;
+  }
+
```

Comment:
A check if `log_forwarding_labels_exclude` is an empty string (most common case) would allow an early exit from the function.

Reply:
If the exclude rule is empty we need to copy all the labels into the log labels structure so I don't see how we could exit early.

Reply:
Why do we need to copy at that point?
Can't we just return labels?

Reply:
Ok I misunderstood what you mean by "early exit".  Yes it would in theory be possible to reduce the work done in the case of no excludes.

Reply:
Multiple paths to a refactor being considered for future efforts.